modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/CefParser.java

```diff
@@ -0,0 +1,459 @@
+    );
+
+    private final List<ChronoField> CHRONO_FIELDS = List.of(
+        NANO_OF_SECOND,
+        SECOND_OF_DAY,
+        MINUTE_OF_DAY,
+        HOUR_OF_DAY,
+        DAY_OF_MONTH,
+        MONTH_OF_YEAR
+    );
+
+    CEFEvent process(String cefString) {
+        List<String> headers = new ArrayList<>();
+        Matcher matcher = HEADER_NEXT_FIELD_PATTERN.matcher(cefString);
+        int extensionStart = 0;
+
+        for (int i = 0; i < 7 && matcher.find(); i++) {
+            String field = matcher.group(1);
+            field = HEADER_ESCAPE_CAPTURE.matcher(field).replaceAll("$1");
+            headers.add(field);
+            extensionStart = matcher.end();
+        }
+
+        if (headers.isEmpty() == false && headers.getFirst().startsWith("CEF:")) {
+            CEFEvent event = new CEFEvent();
+            // Add error message if there are not enough header fields
+            if (headers.size() != 7) {
+                event.addRootMapping("error.message", new HashSet<>(ERROR_MESSAGE_INCOMPLETE_CEF_HEADER));
+            }
+            processHeaders(headers, event);
+            processExtensions(cefString, extensionStart, event);
+            return event;
+        } else {
+            throw new IllegalArgumentException("Invalid CEF format");
+        }
+    }
+
+    private static void processHeaders(List<String> headers, CEFEvent event) {
+        for (int i = 0; i < headers.size(); i++) {
+            final String value = headers.get(i);
+            switch (i) {
+                case 0 -> event.addCefMapping("version", value.substring(4));
+                case 1 -> {
+                    event.addCefMapping("device.vendor", value);
+                    event.addRootMapping("observer.vendor", value);
+                }
+                case 2 -> {
+                    event.addCefMapping("device.product", value);
+                    event.addRootMapping("observer.product", value);
+                }
+                case 3 -> {
+                    event.addCefMapping("device.version", value);
+                    event.addRootMapping("observer.version", value);
+                }
+                case 4 -> {
+                    event.addCefMapping("device.event_class_id", value);
+                    event.addRootMapping("event.code", value);
+                }
+                case 5 -> event.addCefMapping("name", value);
+                case 6 -> event.addCefMapping("severity", value);
+            }
+        }
+    }
+
+    private void processExtensions(String cefString, int extensionStart, CEFEvent event) {
+        String extensionString = cefString.substring(extensionStart);
+        Map<String, String> extensions = parseExtensions(extensionString);
+
+        // Cleanup empty values in extensions
+        if (removeEmptyValue) {
+            removeEmptyValue(extensions);
+        }
+
+        // Translate extensions to possible ECS fields
+        Map<String, Object> translatedFields = extensions.entrySet()
```

Comment:
Please rewrite this streaming block into ordinary banging on collections -- this is code is too hot for streaming (also there are some wasteful things you're doing in the streaming version that will drop out trivially in the non-streaming version.